pkg/oauth2/login.go

```diff
@@ -0,0 +1,149 @@
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package oauth2
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/spf13/pflag"
+	o "github.com/streamnative/pulsarctl/pkg/auth/oauth2"
+	store2 "github.com/streamnative/pulsarctl/pkg/auth/oauth2/store"
+	"github.com/streamnative/pulsarctl/pkg/cmdutils"
+	"github.com/streamnative/pulsarctl/pkg/oauth2/os"
+)
+
+func loginCmd(vc *cmdutils.VerbCmd) {
+	desc := cmdutils.LongDescription{}
+	desc.CommandUsedFor = "This command is used for oauth2 user login."
+	desc.CommandPermission = "This command doesn't need pulsar permissions."
+
+	var examples []cmdutils.Example
+	login := cmdutils.Example{
+		Desc:    "Login as a oauth2 user",
+		Command: "pulsarctl oauth2 login --issuer-endpoint (issuer) --audience (audience) --client-id (client-id)",
+	}
+	examples = append(examples, login)
+	desc.CommandExamples = examples
+
+	vc.SetDescription(
+		"login",
+		"Login to oauth2 server",
+		desc.ToString(),
+		desc.ExampleToString(),
+		"login")
+
+	var issuerEndpoint, clientID, audience, keyFile string
+
+	vc.SetRunFunc(func() error {
+		return doLogin(vc, issuerEndpoint, clientID, audience, false)
+	})
+
+	vc.FlagSetGroup.InFlagSet("Oauth2 login", func(set *pflag.FlagSet) {
+		set.StringVarP(&issuerEndpoint, "issuer-endpoint", "i", "",
+			"The OAuth 2.0 issuer endpoint")
+		set.StringVarP(&audience, "audience", "a", "",
+			"The audience identifier for the API server")
+		set.StringVarP(&clientID, "client-id", "c", "",
+			"The client ID to user for authorization grants")
+		set.StringVarP(&keyFile, "key-file", "k", "",
+			"Path to the private key file")
+	})
+}
+
+func doLogin(vc *cmdutils.VerbCmd, issuerEndpoint, clientID, audience string, noRefresh bool) error {
+	if issuerEndpoint == "" || clientID == "" || audience == "" {
+		return errors.New("the arguments issuer-endpoint, client-id, audience can not be empty")
+	}
+	issuer := o.Issuer{
+		IssuerEndpoint: issuerEndpoint,
+		ClientID:       clientID,
+		Audience:       audience,
+	}
+
+	options := o.DeviceCodeFlowOptions{
+		AdditionalScopes: nil,
+		AllowRefresh:     noRefresh,
+	}
+
+	prompt := NewPrompt(false)
+	flow, err := o.NewDefaultDeviceCodeFlow(issuer, options, prompt.Prompt)
+	if err != nil {
+		return errors.New("configuration error: unable to use device code flow: " + err.Error())
+	}
+	grant, err := flow.Authorize()
+	if err != nil {
+		return errors.New("login failed: " + err.Error())
+	}
+
+	fmt.Println(grant.Token.AccessToken)
+
+	store, err := store2.MakeKeyringStore()
```

Comment:
We should parameterize `MakeKeyringStore` to take the keyring name, since it is probably using `snctl` as a const.  here we would want `pulsarctl`.

Reply:
This is the implementation of the MakeKeyringStore: https://github.com/zymap/pulsarctl/pull/8/files#diff-f5d260c645f08004f2e0bbd021abb0d3R27. I change the keyring name to the pulsarctl.